[#2691] Track IP addresses associated with User signins

garethrees opened the pull request:
If a retention period is configured, create a record associated with the
user of the signin and the IP address used to do so.

A rake/cron task is provided to purge records older than the retention
period.

A version of https://github.com/mysociety/alaveteli/issues/2691

---

View/search all recorded sign ins:

<img width="1350" alt="Screenshot 2022-02-25 at 21 36 47" src="https://user-images.githubusercontent.com/282788/155806691-dba2e521-2c7c-4d2c-ae4f-da844772355f.png">

View a given user's sign ins. Check how many others have signed in with the IP:

<img width="1350" alt="Screenshot 2022-02-25 at 21 36 27" src="https://user-images.githubusercontent.com/282788/155806671-9410fe7e-af72-441b-bcda-415ceb558190.png">

Click the IP address to go through to the main listing filtered by that IP (or country, added after screenshot):

<img width="1350" alt="Screenshot 2022-02-25 at 21 36 38" src="https://user-images.githubusercontent.com/282788/155806686-243d2f9b-3124-476f-86e6-31f1dcad1b3b.png">

Filter by partial IP:

<img width="1350" alt="Screenshot 2022-02-25 at 21 55 32" src="https://user-images.githubusercontent.com/282788/155809296-a20212e5-f234-4bc1-9ed8-b81cfb014b1a.png">

Filter by country:

<img width="1350" alt="Screenshot 2022-02-25 at 21 55 23" src="https://user-images.githubusercontent.com/282788/155809302-75d38c21-2bfa-4204-91e7-c6167710f2ee.png">

garethrees commented on app/controllers/admin/users/sign_ins_controller.rb:
```diff
@@ -0,0 +1,17 @@
+class Admin::Users::SignInsController < AdminController
```
Will add class-level docs on merge.

garethrees commented on config/routes.rb:
```diff
@@ -652,6 +652,14 @@
   ####
 
+  #### Admin::Users::SignIns controller
+  namespace :admin do
```
Couldn't quite decide on how to route this.

Alternative was…

```ruby
namespace :admin do
  namespace: :users do
    resources :sign_ins, only: [:index] #=> /admin/users/sign_ins
  end
end
```

…but wanted to avoid nesting hell. Could possibly do the above and use shallow routes if more actions added. Open to suggestion.

garethrees commented on app/views/admin/users/_sign_in_table.html.erb:
```diff
@@ -10,6 +10,10 @@
       <td><%= admin_date(sign_in.created_at, ago_only: true) %></td>
+
+      <td>
+        <strong><%= sign_in.other_users.size %></strong> others using this IP
```
Probably okay for now, but I think we'll want to handle [Apple Private Relay](https://developer.apple.com/support/prepare-your-network-for-icloud-private-relay/). Probably annotate the stored country with a note so that a high count + APR country wouldn't create suspicion?

garethrees replied:
Might want to add some conditional for when we have no recorded sign ins

gbp replied:
```suggestion
        <strong><%= pluralize sign_in.other_users.size, 'other' %></strong> using this IP
```

garethrees replied:
> `pluralize sign_in.other_users.size`

I thought about this, but I wanted only the number to be `<strong>`.

Could enhance further with only bolding if number is > 1, but not clear that's useful yet.

garethrees replied:
> Might want to add some conditional for when we have no recorded sign ins

Noted elsewhere.

> Need to return distinct users

Noted elsewhere.

gbp commented on app/models/user/sign_in.rb:
```diff
@@ -0,0 +1,47 @@
+  end
+
+  def other_users
+    self.class.where(ip: ip).where.not(user_id: user_id)
```
Should return distinct User instances not User::SIgnIn instances.

```suggestion
    User.distinct.joins(:sign_ins).
      where(user_sign_ins: { ip: ip }).
      where.not(user_sign_ins: { user_id: user_id })
```

gbp replied:
Or:

```suggestion
    User.distinct.joins(:sign_ins).
      where(user_sign_ins: { ip: ip }).
      where.not(id: user_id)
```

garethrees replied:
Also add a failing spec

gbp commented on app/controllers/admin/users/sign_ins_controller.rb:
```diff
@@ -0,0 +1,17 @@
+    @title = 'Listing user sign ins'
+    @query = params[:query]
+
+    sign_ins =
+      if @query
+        User::SignIn.search(@query)
+      else
+        User::SignIn
+      end
```
Reduce to two lines:

```suggestion
    sign_ins = User::SignIn.search(@query) if @query
    sign_ins ||= User::SignIn
```


or:

```suggestion
    sign_ins = User::SignIn
    sign_ins = sign_ins.search(@query) if @query
```

garethrees replied:
Prefer the latter as it's clearer it's an additive filter.

gbp commented on db/migrate/20220225214524_add_country_to_user_sign_ins.rb:
```diff
@@ -0,0 +1,5 @@
+class AddCountryToUserSignIns < ActiveRecord::Migration[6.1]
+  def change
+    add_column :user_sign_ins, :country, :string
```
Combine into one migration?

gbp reviewed (changes requested):
This is good, I think there is still some work to do tho.

Does this work with both IPv4 and IPv6 addresses. I think we also need to consider what we do when a user is closed and anonymised.

gbp commented on app/models/user/sign_in.rb:
```diff
@@ -0,0 +1,47 @@
+  before_create :create?
+
+  def self.purge
+    where('created_at < ?', retention_days.days.ago).destroy_all
```
```suggestion
    where(created_at: (...retention_days.days.ago)).destroy_all
```

garethrees replied:
To me the former is much clearer. Just resolving for now but happy for a PR later to change.

gbp commented:
How about making this more generic and recording IP on sign ins and sign ups?

Could then link the "rate limited signup from" email to the index action with IP already filtered.

garethrees commented on app/views/admin/users/_sign_in_table.html.erb:
```diff
@@ -0,0 +1,9 @@
+<table class="table table-condensed table-striped">
```
Add "none yet" conditional

garethrees commented on app/views/admin_user/show.html.erb:
```diff
@@ -156,6 +156,19 @@
 
+<h2>Sign Ins</h2>
+
+<table class="table table-condensed table-striped">
```
Add "none yet" conditional in this commit

garethrees commented:
> Does this work with both IPv4 and IPv6 addresses.

I've now used postgres inet type and added a basic test, so _should_ do, though I don't know much about IPv6 tbh.

> I think we also need to consider what we do when a user is closed and anonymised.

Good point. I think the regular purge would have been fine, but also added on-demand purging via `User#close_and_anonymise` in 97c6d71815f14af4f7f320d66e9c6e6a295384b4.

> How about making this more generic and recording IP on sign ins and sign ups? Could then link the "rate limited signup from" email to the index action with IP already filtered.

I did think about that, but this is the more immediate problem. To consider in relation to #6931.

garethrees commented:
Couple of things left to do, but in order to do so I'll want to squash these commits:

- [ ] Move the addition of `default_scope` to the previous (first) commit
- [ ] Add "none yet" conditional to the sign in table
- [ ] Squash eab805a and 73e939d into a single commit (~ "Render user sign ins in the admin interface")
- [ ] Handle `USER_SIGN_IN_ACTIVITY_RETENTION_DAYS == 0`. Instead of hiding, I'll use a notice saying "sign in tracking is not configured; see general.yml", or something to that effect.

garethrees commented:
@gbp a quick pair of eyes over the fixups would be great before I squash them down to complete the [remaining tasks](https://github.com/mysociety/alaveteli/pull/6822#issuecomment-1112092653).

gbp commented:
> @gbp a quick pair of eyes over the fixups would be great before I squash them down to complete the [remaining tasks](https://github.com/mysociety/alaveteli/pull/6822#issuecomment-1112092653).

Fixups look good me, feel free to squash and rebase

garethrees commented:
Think this is good to go now 🥳

gbp reviewed (approved):
LGTM

garethrees commented:
Linking to https://github.com/mysociety/alaveteli/pull/7592.
